### PR TITLE
Bolt EUV: Non-linear torque function

### PR DESCRIPTION
**Pull request by nuwandavek**
# Changes
- Describe torque vs lateral acceleration with a non-linear curve (credit for insight, inspiration and extensive testing: @twilsonco)  
- Refactor the way torque(lataccel error) is computed

ToDo: 
- [x] Try to fit a non-linear relationship between steer-torque and lateral acceleration
- [x] Bugfix: `f(x+y)!= f(x) + f(y)` in nonlinear functions
- [x] Refs
---
- [ ] ~Try to make a more generic GM fit (consider slope at 0)~ Explore `tanh` once this is merged
- [ ] ~Remove low-speed-factor?~ Explore in conjunction with PID vs PI controller

### CODE REVIEW

**Comment by Verylukyguy** [comment]
I’d be happy to help test this on my vehicle; the 2018 GMC Acadia has a similar non-linear function.

**Comment by nuwandavek** [comment]
> I’d be happy to help test this on my vehicle; the 2018 GMC Acadia has a similar non-linear function.

As soon as I merge this, will make a generic GM function.

File: selfdrive/car/gm/interface.py [hunk 1]
```diff
@@ -1,13 +1,15 @@
 #!/usr/bin/env python3
 from cereal import car
-from math import fabs
+from math import fabs, exp  # , tanh, erf
```
Comment (adeebshihadeh) [inline selfdrive/car/gm/interface.py:3]: what's the comment for?

File: selfdrive/car/gm/interface.py [hunk 2]
```diff
@@ -44,6 +46,34 @@ def get_steer_feedforward_function(self):
...
+    friction = friction_interp if friction_compensation else 0.0
+
+    def sig(val):
+      return 1 / (1 + exp(-val)) - 0.5
+
+    # The "lat_accel vs torque" relationship is assumed to be the sum of "sigmoid + linear" curves
+    # An important thing to consider is that the slope at 0 should be > 0 (ideally >1)
+    # This has big effect on the stability about 0 (noise when going straight)
+    # ToDo: To generalize to other GMs, explore tanh function as the nonlinear
+    a, b, c, _ = [2.6531724862969748, 1.0, 0.1919764879840985, 0.009054123646805178]  # weights computed offline
```
Comment (adeebshihadeh) [inline selfdrive/car/gm/interface.py:65]: what's the underscore value?

File: selfdrive/car/gm/interface.py [hunk 3]
```diff
@@ -44,6 +46,34 @@ def get_steer_feedforward_function(self):
...
+    steer_troque_fn = sig
```
Comment (adeebshihadeh) [inline selfdrive/car/gm/interface.py:66]:
```suggestion
    steer_torque_fn = sig
```

File: selfdrive/car/gm/interface.py (hunk 2 above)
Comment (nuwandavek) [reply selfdrive/car/gm/interface.py:65]: It is the constant, which is useful to completely describe the curve, but not necessary to be used in the function `steer_from_lataccel` fn

File: selfdrive/controls/lib/drive_helpers.py [hunk 4]
```diff
@@ -190,3 +190,13 @@ def get_lag_adjusted_curvature(CP, v_ego, psis, curvatures, curvature_rates):
                                 current_curvature_desired + max_curvature_rate * DT_MDL)
 
   return safe_desired_curvature, safe_desired_curvature_rate
+
+
+def get_friction(lateral_accel_error, lateral_accel_deadzone, friction_threshold, torque_params, friction_compensation):
```
Comment (adeebshihadeh) [inline selfdrive/controls/lib/drive_helpers.py:195]: can you add type hints to this function?

**Review by adeebshihadeh — APPROVED** [review]

**Comment by sshane** [comment]
🥳

**Comment by MoreTore** [comment]
The way you guys are handling the vehicle-specific non-linear torque makes it difficult to merge other non-linear vehicles like mine :(

**Pull request merged**